api/src/main/java/io/grpc/BufferDrainable.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import java.io.IOException;
+import java.nio.BufferOverflowException;
+import java.nio.ByteBuffer;
+
+/**
+ * Extension to an {@link java.io.InputStream} or alike by adding a method that
+ * allows transferring directly to an underlying {@link ByteBuffer}
+ */
+public interface BufferDrainable {
+
+  /**
+   * Transfers all content to a target {@link ByteBuffer}
+   *
+   * @param dest the destination buffer to receive the bytes.
+   * @throws BufferOverflowException if destination buffer has insufficient remaining bytes
+   */
+  int drainTo(ByteBuffer dest) throws IOException;
```

Comment:
I don't disagree and this is actually what I started with for exactly the reasons you described, but I was thinking that the target impl doesn't actually work that way so this has the side effect that it will only be used when beneficial. And at least it's kind of similar to both `ReadableBuffer#readBytes(ByteBuffer)` which requires the target buffer to be large enough to hold the entire source, and the `Drainable` interface which has to drain the entire InputStream.

Having `ProtoInputStream` fall back to use `ByteArrayInputStream` in the case that the provided `ByteBuffer` isn't big enough would be worse than the current path, and this would not be obvious from the interface. WDYT?

It's also why I changed the name to `drainTo`, agree `read` more appropriate if just filled the `ByteBuffer`.

Reply:
> I was thinking that the target impl doesn't actually work that way so this has the side effect that it will only be used when beneficial

"Beneficial" is in the eye of the beholder. Yes, with the current code that is true. If this was in an API we could readily change, I'd agree. But since this is a public API that 1) will last a long time and 2) have multiple users I'd much rather favor the more generic method definition. I'm optimizing the API for longevity while making sure it allows optimizing the implementation.

As things change, my preference order would be:
1. Any old code path is fast
2. Any old code path works
3. Any old code path fails (and this one is not actually acceptable)

Since `drainTo(ByteBuffer)` would _require_ `KnownLength`, I'm not as fond of it. It could be useful to drain a compressed message to a `ByteBuffer` or some such.

Interfaces generally don't define guaranteed performance levels. List.get() may be fast or slow. We will consider performance when making changes, but at times we may need to hurt performance of one particular method for some other goal. So I'm okay with the varying level of performance. If you are benchmarking then you don't care about performance, and such a regression would show up on a benchmark.

Reply:
> If this was in an API we could readily change, I'd agree.

How about just moving it to `internal` package then (alongside `WritableBuffer`)?

Probably this is clear but I meant that performance-wise the only implementation of this would actually involve _more_ copying than less in the case that the stream didn't fit in the buffer i.e. worse than it is now and having the opposite than intended effect of this PR.